### PR TITLE
Extend library [WIP]

### PR DESCRIPTION
**Pull request by kalbasit**

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+func Open(path string, mode DBMode) (*DB, error) {
+	cpath := C.CString(path)
+	defer C.free(unsafe.Pointer(cpath))
+
+	cmode := C.notmuch_database_mode_t(mode)
+	db := &DB{}
+	cdb := (**C.notmuch_database_t)(unsafe.Pointer(&db.cptr))
+	cerr := C.notmuch_database_open(cpath, cmode, cdb)
+	runtime.SetFinalizer(db, func(db *DB) {
+		C.notmuch_database_destroy(db.toC())
```
Comment (kalbasit) [inline db.go:60]: about https://travis-ci.org/zenhack/go.notmuch/jobs/82026238#L1819 not sure why this is blowing up, I confirmed that calling `db.Close` multiple times does not panic. @zenhack thoughts?

File: query.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		return nil, err
+	}
+	runtime.SetFinalizer(threads, func(t *Threads) {
+		C.notmuch_threads_destroy(t.toC())
+	})
+	return threads, nil
+}
+
+// CountThreads returns the number of messages for the current query.
+func (q *Query) CountThreads() int {
```
Comment (zenhack) [inline query.go:39]: As implemented this is identical to CountMessages; I assume that's a mistake.

Comment (kalbasit) [inline query.go:39]: yep it is. Nice catch!

File: db_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	defer func() {
+		os.RemoveAll(tmpDbPath)
+	}()
+
+	db, err := Create(tmp)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if got := db.Version(); got == 0 {
+		t.Errorf("db.Version(): expected to return something")
```
Comment (zenhack) [inline db_test.go]: Might be good to document what; 0 is _something_.

File: thread.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// GetID returns the ID of the thread.
+func (t *Thread) GetID() string {
+	return t.id
+}
+
+// Next retrieves the next message from the thread. Next returns true if a message
+// was successfully retrieved.
+func (t *Thread) Next(m *Message) {
```
Comment (zenhack) [inline thread.go]: So, obviously this is unimplemented and the signature disagrees with the comment. Also, this doesn't appear to wrap a notmuch function.

I'd say let's just defer adding additional wrappers to another pr, but my $0.02: I'd have a method on `*Thread` that returns a `*Messages` instead.

File: db_test.go (hunk 3 above)
Comment (kalbasit) [inline db_test.go]: done

File: thread.go (hunk 4 above)
Comment (kalbasit) [inline thread.go]: oops that shouldn't have made it here. Nice catch.

**Comment by zenhack** [comment]
General comment: Splitting things up into files by type is a good idea, but if you're going to do that you should go all the way with it. You've got a number of things defined in the wrong file for this scheme. Also, status is weirdly split up (type is defined in db.go, statusErr in utils.go).

**Comment by kalbasit** [comment]
I moved the status to utils.go (nice catch) as for the other things. I tried to follow notmuch's file structure so [`notmuch_query_search_threads_st`](https://github.com/notmuch/notmuch/blob/93ee4faa4d60e78c8cfeb95580919cac07ac0eca/lib/query.cc#L446) which is used by `notmuch.(*DB).Query.Threads()` is defined in `lib/query.cc` so I kept it in the same file. Please add comments about functions you'd like moved.

File: db.go (hunk 1 above)
Comment (kalbasit) [inline db.go:60]: funny that it's always failing on Go 1.4 and passing on Go 1.5. @zenhack are we planning to support Go 1.4?

**Comment by zenhack** [comment]
My comment on IRC:

> It seems inconsistent to have both of those in query.go when you have a separate file for db. Maybe less 
> so when they're functions and it's not totally defined what they're "methods" on. I don't know that basing 
> the file layout on upstream makes any more sense than doing so with the API (which was the first thing 
> you thought needed improvement).

I think these are reasonable guidlines:
- One file per type
- Methods go in the same file as their reciever. (so e.g. NewQuery goes in db.go)

This probably implies adding threads.go/messages.go

**Comment by zenhack** [comment]
There's some inconsistency with how you're handling the thread id. You're defining a field for it in `Thread` and returning that field in `Thread.GetID()`, but you're never actually populating it. My thought would be to just get rid of the field and call `C.notmuch_thread_get_thread_id` from `GetID`.

**Comment by kalbasit** [comment]
So the ID is actually being populated through `(ts *Threads) get() *Thread` but I do feel it's cleaner to use `C. notmuch_thread_get_thread_id`. done

**Comment by zenhack** [comment]
Okay, I'm happy. Travis seems to be as well; merging.

File: db.go (hunk 1 above)
Comment (zenhack) [inline db.go:60]: It seems to have fixed itself, but for the record, no I don't care about go 1.4

**Pull request merged**